docs: add extended passport support (#142)
decentraland/adr: content/ADR-142-extended-passport-support.md

---
adr: 142
date: 2022-11-25
title: Technical Assessment for the Extended Passport
authors:
  - manumena
status: Draft
type: RFC
spdx-license: CC0-1.0

> redirect_from:
>   - /rfc/RFC-1

Reviewer: ```suggestion
```

---

> 

Reviewer 2: ## Abstract

The Extended Passport view for Decentraland's Explorer requires storing and retrieving data from profiles to display NFT collectibles, blocking users, adding friends, and initiating chats. This document outlines the necessary requirements and performance considerations for the Catalyst, including the introduction of a user through the "description" field in profiles, the use of the "blocked" field to display blocked users, and the creation of new endpoints in lambdas for each type of item. The document also notes that the Explorer team should implement lazy loading and pagination to improve performance and scalability, as some users may have more than 1,000 NFTs. The document also provides details on the various endpoints that will be used to retrieve this data.

Author: This is awesome Mariano, thank you very much

Reviewer 2: Don't thank me... thank https://chat.openai.com/chat 😉

Author: :O

## Need

A New Passport view is being developed by the Explorer Team.

This view requires storing some information in the profile entity and retrieving other data (such as collectibles) to be displayed. It also allows blocking users, adding a user as a friend or initiating a chat.

## Requirements
The applicable requirements from the Catalyst point of view:

1. Intro of a user:

> Use the "description" field from profiles to use in the "Intro" section.

Reviewer 2: ```suggestion
use the "description" field from profiles to use in the "Intro" section.
```

1. Block/Unblock users:

> Use the "blocked" field from profiles to see/report blocked users.

Reviewer 2: ```suggestion
use the "blocked" field from profiles to see/report blocked users.
```

1. Equipped wearables:

> Take the information from the profile. There are the equipped wearables, emotes and the current name.

Reviewer 2: ```suggestion
take the information from the profile. There are the equipped wearables, emotes and DCL name.
```

> 1. Items: wearables/emotes/names/lands: New endpoints will be created in [lambdos](https://github.com/decentraland/lambdos) for every type of item.

Reviewer 2: ```suggestion
1. Items: wearables/emotes/names/lands: new endpoints will be created in [lambdos](https://github.com/decentraland/lambdos) for each type of item.
```

> 1. Guest: Passport for guest users will show only name and description, there is nothing to do for guests from the Catalyst end.

Reviewer 2: ```suggestion
1. Guest: passport for guest users will show only name and description, there is nothing to do for guests from the Catalyst end.
```

> 2. Social network links: This information is not currently available. It will be displayed in the profile under `socialLinks`.

Reviewer 2: ```suggestion
1. Social network links: this information is not currently available. It will be displayed in the profile under `socialLinks`.
```

1. `lambdos/nfts/wearables/:address`
1. `lambdos/nfts/emotes/:address`
1. `lambdos/nfts/lands/:address`
1. `lambdos/nfts/names/:address`

Once this information is obtained, to display wearables and emotes a few extra requests will be needed for each emote/wearable to download their image/thumbnail. This does not scale because some users have more than 1k NFTs. So the Explorer should implement a lazy loading mechanism (for image/thumbnail) as the user browses the collectibles.

All endpoints except `/profiles` are expensive requests. Internally, endpoints 2 to 5 make queries to TheGraph to check owned NFTs urns and then to the Content Server to get the definitions. Currently, endpoints 2 and 3 are not paginated, so making a query for a user that owns more than 1k wearables would be considerably heavy. Also, sometimes queries to TheGraph are slow. So to improve this situation the endpoints from 2 to 5 should be paginated.

### Pagination

The page number and size will be dynamic for every endpoint, received as query parameters. 

- `pageSize: number`
- `pageNum: number`

**Pagination will be optional**: if any of this parameters is not present, the response won't be paginated. 

> Items displayed in each page will be ordered descending by the `transferredAt` field of the `NFT` entity (representing the date in which the NFT was received, currently in development: https://github.com/decentraland/marketplace/issues/1092).

Reviewer 2: ```suggestion
Items displayed in each page will be ordered descending by the `transferredAt` field of the `NFT` entity (representing the date in which the NFT was received, see implementation [here](https://github.com/decentraland/marketplace/issues/1092)).
```

### Wearables

The wearables are searched in 2 different collections: **ethereum** and **matic**. That involves doing 2 different queries to TheGraph.

> To be able to return a paginated and sorted response, a full query is made for both, then that result is merged, sorted and stored in a LRU cache where the key is the address. Pages are served upon that stored data.

Reviewer 2: ```suggestion
To be able to return a paginated and sorted response, a full query is made for both, then that results are merged into a single one, then sorted and stored in a LRU cache where the key is the address. Paginated requests are resolved using the cached data.
```

---
adr: 142
date: 2022-11-25
title: Technical Assessment for the Extended Passport
authors:
  - manumena

> status: LastCall
> type: Standars Track

Reviewer 2: ```suggestion
status: LastCall
type: Standards Track
```
